dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,272 @@
+        }
+
+        co_return nullptr;
+    }
+
+    PushNotificationRegistrationToken PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    {
+        THROW_HR_IF_NULL(E_INVALIDARG, details);
+        THROW_HR_IF(E_INVALIDARG, details.TaskClsid() == winrt::guid());
+        THROW_HR_IF(E_INVALIDARG, !(WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::PushTrigger) || WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::ComActivator)));
+
+        GUID taskClsid = details.TaskClsid();
+        DWORD cookie = 0;
+        BackgroundTaskRegistration registeredTask = nullptr;
+        BackgroundTaskBuilder builder = nullptr;
+
+        if (WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::PushTrigger))
+        {
+            winrt::hstring taskClsidStr = winrt::to_hstring(taskClsid);
+            auto tasks = BackgroundTaskRegistration::AllTasks();
+            bool taskRegistered = std::any_of(begin(tasks), end(tasks),
+            [&](auto&& task)
+            {
+                std::wstring backgroundTaskNameStr = task.Value().Name().c_str();
+                if (backgroundTaskNameStr.find(backgroundTaskName) != std::wstring::npos)
+                {
+                    if (backgroundTaskNameStr.find(taskClsidStr) != std::wstring::npos)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
+                    }
+                };
+                return false;
+            });
+
+            if (!taskRegistered)
+            {
+                builder = BackgroundTaskBuilder();
+                builder.Name(backgroundTaskName + taskClsidStr.c_str());
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                if (IsPackagedProcess() && builder5)
+                {
+                    builder5.SetTaskEntryPointClsid(taskClsid);
+                    winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                    for (auto condition : conditions)
+                    {
+                        builder.AddCondition(condition);
+                    }
+                }
+                else
+                {
+                    throw winrt::hresult_not_implemented();
+                }
+            }
+        }
+
+        if (WI_IsFlagSet(details.Option(), PushNotificationRegistrationOption::ComActivator))
+        {
+            {
+                auto lock = g_lock.lock();
+                // Define handle that will be set during background task execution
+                g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
```

Comment:
>g_waitHandleForArgs [](http://example.com/codeflow?start=16&length=19)

I had a hard time matching the lifecycle of this event and the args; should this event be pre-constructed as a global, rather than set here?